### DISCUSSION TITLE
ActiveAdmin 3.x stylesheet warnings: lighten() and darken() deprecated

### DISCUSSION THREAD

**Discussion by mattbrictson** [activeadmin/activeadmin#8492] [Help]
I am using ActiveAdmin 3.2.5 with dart sass 1.79.3.

### Expected behavior

Compiling assets should complete without warnings.

### Actual behavior

There are some new deprecation warnings introduced by dart sass 1.79.0. The color functions `lighten` and `darken` are deprecated. ActiveAdmin 3.x uses these functions in several places in its scss files. Here is one example:

```
DEPRECATION WARNING: lighten() is deprecated. Suggestions:

color.scale($color, $lightness: 7.5%)
color.adjust($color, $lightness: 5%)

More info: https://sass-lang.com/d/color-functions

   ╷
10 │   @include gradient(lighten($primary-color, 5%), darken($primary-color, 7%));
   │                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   ╵
    node_modules/@activeadmin/activeadmin/src/scss/mixins/_gradients.scss 10:21  primary-gradient()
    node_modules/@activeadmin/activeadmin/src/scss/_header.scss 3:3              @import
    node_modules/@activeadmin/activeadmin/src/scss/_base.scss 8:11               @import
    app/frontend/entrypoints/active_admin.scss 20:9                              root stylesheet
```

### How to reproduce

Generate a blank Rails 7.2 app:

```
rails new sass-repro
```

Install dart sass:

```
./bin/bundle add dartsass-rails
./bin/rails dartsass:install
```

Install ActiveAdmin:

```
./bin/bundle add activeadmin
./bin/rails g active_admin:install --skip-users
```

For a simple reproduction, replace the Rails default `application.scss` with ActiveAdmin's:

```
cp app/assets/stylesheets/active_admin.scss app/assets/stylesheets/application.scss
```

Compile assets:

```
./bin/rails assets:precompile
```

Produces:

```
DEPRECATION WARNING: lighten() is deprecated
...
```

**Comment by mattbrictson** [comment]
In my own code, I have been solving these deprecation warnings as follows.

Before:

```
lighten($primary-color, 5%)
```

After:

```
@use "sass:color";

color.adjust($primary-color, $lightness: 5%, $space: hsl)
```

This will supposedly produce the identical color output as the deprecated code, according to https://sass-lang.com/documentation/modules/color/#lighten

For `darken`, I am following the same pattern, with a negative number for the `$lightness` value. E.g.:

```
// darken 7%
color.adjust($primary-color, $lightness: -7%, $space: hsl)
```

**Comment by mattbrictson** [comment]
One potential drawback of fixing these deprecation warnings is that I'm not sure that Ruby sass understands the `@use` syntax. So Rails apps that are using the Ruby `sass` gem (as opposed to `dartsass-rails` or its Node equivalent) may break. The sass gem hasn't been released in over 8 years, so maybe this is not a big concern.

_Edit: Apps that use `sassc`, which is more recent (4 years old), may also break, since I believe it is based on `libsass` and doesn't support newer scss syntax._

**Reply by javierjulio** [reply]
That is a concern as maintaining `sassc` support for v3 is important for now. Although that doesn't matter on v4. No interest in supporting dart sass deprecations as they are easy to ignore. I would consider an update if there is some way [we could conditionalize this for dart sass and sassc](https://stackoverflow.com/a/20712730). Let's see if we can get others to suggest some ideas as well. Thank you.

**Reply by mattbrictson** [reply]
I found that warnings from ActiveAdmin's stylesheets can be silenced using the `quietDeps` option.

https://sass-lang.com/documentation/breaking-changes/color-functions/#silencing-deprecations-in-dependencies

I use Vite, so I accomplished this by modifying my `vite.config.ts` as follows:

```diff
--- vite.config.ts
+++ vite.config.ts
@@ -4,6 +4,13 @@ import sassGlobImports from "vite-plugin-sass-glob-import";
 import StimulusHMR from "vite-plugin-stimulus-hmr";
 
 export default defineConfig({
+  css: {
+    preprocessorOptions: {
+      scss: {
+        quietDeps: true,
+      },
+    },
+  },
   plugins: [
     ViteRails({ envVars: { RAILS_ENV: "development" } }),
     sassGlobImports(),
```

**Reply by javierjulio** [reply]
@mattbrictson I think it would be best to resolve this by not using the `lighten` or `darken` functions and avoiding `@use` as well for compatibility. The values should be defined in a variable the user can set and change. What do you think about adding new variables to replace those and adding a conditional where if the variable is undefined [to print a warning explaining that it should be set](https://sass-lang.com/documentation/at-rules/warn/)? This would make it compatible with both sassc and dartsass and would resolve the deprecation warning. It would benefit others as they transition to dartsass while on v3. Would you like to contribute that update?

**Reply by mattbrictson** [reply]
@javierjulio Hmm. I would rather not introduce more warnings. Is it fair to say that users who have already migrated to dart-sass are likely going to be adopters of v4, as well? Whereas users sticking with v3 are likely also going to be using older tech like sassc?

In which case I think keeping the experienced unchanged for v3 users is probably for the best. Users like me who are trying to make v3 work with dart-sass can be expected to have to put in some extra effort and/or deal with warnings.

So maybe we can extract new variables like you are suggesting, but still use the sassc lighten/darken functions for full backwards compatibility, like this:

```scss
$primary-gradient-start: lighten($primary-color, 5%) !default;
$primary-gradient-stop: darken($primary-color, 7%) !default;
```

And then in the documentation, we can tell users of dart-sass that they will need to manually define these variables to avoid deprecation warnings. If they define the variables themselves, then the lighten/darken functions will be skipped, due to the fact that `!default` variable assignments are skipped if the variable already has a value.

What do you think?

**Reply by javierjulio** [reply]
@mattbrictson I think that's great. Thank you. I agree, I wouldn't want more deprecation warnings but my thinking was to just notify users affected so they know what to do to resolve it. Calling it out in the changelog instead would help as a non-warning alternative.

But what I didn't know is what you mentioned about `!default` variable assignments as I figured if it still contained deprecated code that it would warn. If you know that it will not emit any warning on Dart Sass for that case, I agree that's best. Please feel free to open a PR so that can be fixed for you and others. We'll cut another v3 release.

Do you know of any other Dart Sass warnings currently being emitted on ActiveAdmin v3 with the latest? I'm hopeful it's only the `@import` deprecation ([source](https://sass-lang.com/blog/import-is-deprecated/)) which won't be removed anyway in v2. As they suggest it would be best to just ignore that warning.

**Reply by mattbrictson** [reply]
@javierjulio here is the test I did:

```scss
$primary-color: #5E6469 !default;

// Explicit assignment causes the subsequent default assignment to be skipped
$primary-gradient-start: #6a7176;

// No deprecation warning triggered since the assignment is skipped
$primary-gradient-start: lighten($primary-color, 5%) !default;

body {
  color: $primary-gradient-start;
}
```

Using the latest dart-sass (`sass@1.80.6` Node package), there are no warnings. So it seems like this approach will work.

Unfortunately, I don't have much bandwidth to contribute a PR right now, but I will try to come back to this later.
